rfd/0175-spiffe-federation.md

```diff
@@ -0,0 +1,429 @@
+Auth Server and there is no need for the user to modify it.
+
+#### Syncing Trust Bundles.
+
+A background task will be introduced to the Teleport Auth Server:
+`sync-spiffe-federation`. This will run on an elected Teleport Auth Server,
+and periodically fetch the trust bundle from the configured Bundle Endpoint.
+This will be written into `status.current_bundle`.
+
+The `current_bundle_refresh_hint` and `current_bundle_synced_at` fields will be
+used to determine when the next fetch should occur. If the `current_bundle` is
+older than the `current_bundle_refresh_hint` from the `current_bundle_synced_at`,
+a new fetch will be initiated.  If unspecified, it will be assumed that the
+bundle should be refreshed every 60 minutes.
+
+In addition, this task will watch the `SPIFFEFederation` resources and
+immediately fetch the trust bundle when a new resource is created or updated.
+
+#### `tbot`
+
+Some minor changes will need to be made to `tbot` to support federation.
+
+When the SPIFFE service starts, it will additionally need to fetch and watch
+the `SPIFFEFederation` resources. When a resource is updated, added or deleted,
+the service will need to update its trust bundle cache and distribute the
+updated trust bundles to any subscribed workloads.
+
+### SPIFFE Joining
+
+SPIFFE joining will be implemented as a new join method.
+
+SPIFFE defines two forms of SVID that are issued to workloads:
+
+- X.509 SVIDs
+- JWT SVIDs
+
+One option would be to validate the identity of the joining node by requiring
+that it offer the X.509 SVID as a client certificate. Unfortunately, this
+has the following problems:
+
+- Building the ability to validate X.509 certificates issued by arbitrary
+  CAs into our TLS servers is complex and error-prone. We risk a compromised
+  CA being able to issue user certificates that would be trusted by Teleport.
+- Traversing TLS-terminating loadbalancers is challenging.
+- Incompatibility with renewal mechanisms that rely on an X.509 client
+  certificate being presented (e.g BotInstance renewals) as only one client
+  certificate can be presented.
+
+JWT SVIDs also come with their own challenges:
+
+- TTL is determined by SPIFFE implementation and could be longer than we'd
+  prefer when it comes to avoiding re-use. This could be mitigated by a 
+  challenge-response nonce based flow using the audience field as we do with
+  other join methods.
+- JWT SVIDs are less common and may not be supported by all SPIFFE
+  implementations.
+
+Instead, we'll use a third option. The joining node will not use the X.509
+certificate as a client certificate, but will instead present it as part of
+the join payload, with the server returning a challenge nonce. The client will
+sign the nonce with the private key corresponding to the X.509 certificate and
+return it to the server. The server will then verify the signature using the
+public key from the X.509 certificate.
```

Comment:
I'm not sure I understand the security reasons listed above as the justification for this method. And I'm struggling to think of a scenario in which this is viable for the client, though.

What client that has received a SVID is going to have access to private key signing abilities? If the client is a `teleport` or `tbot` binary running somewhere, given that they have not joined the cluster yet, they wouldn't be able to ask the cluster to sign something for them. To even get the SVID they would have to have joined the cluster somehow already anyways (a `tsh svid issue` command run locally?) and it feels a bit circular.

If it's some other SPIFFE-compatible binary running somewhere that is getting a SVID we should trust via federation, is the ability to sign a random set of bytes like this likely to be included? It's just getting it's SVID from a Workload API, it's not likely to be able to call out to the signing authority for this purpose, and it itself is a client of that, it's not likely to have signing ability.

I guess my fundamental question is, if we have a trust bundle, and we have received a request to join that comes with a SVID, if we're not able to verify the SVID using the trust bundle then why are we allowing a SPIFFE join method? Doesn't that defeat the point?

If there's a use case or scenario I'm not getting here please let me know!

Reply:
> What client that has received a SVID is going to have access to private key signing abilities?

So the client in this case will be `teleport` or `tbot` - the key thing here is that SPIFFE joining is about allowing them to join with credentials issued by a SPIFFE identity provider that specifically isn't Teleport e.g integrating with other platforms. This is less a feature of Teleport Workload Identity and more a feature of just Teleport generally.

Reply:
Ok. But how are `teleport` and `tbot` going to get access to the private key that signed the SVID? Or am I misunderstanding what the x509 certificate is in the description?

Reply:
They don't require the private key that signed the SVID, only the private key associated with public key that's included on the X.509 SVID - which is provided by the SPIFFE Workload API. Essentially, there's two different things we're trying to validate:

- That the X.509 SVID is legitimate, which we do by checking the signature of the X.509 certificate against the trust bundle. 
- That the client possesses the private key associated with the public key encoded in the X.509 SVID. This is necessary because an X.509 certificate itself is not a "secret", presenting an X.509 certificate is not enough to prove that it's "your" X.509 certificate.

Reply:
Ok, got it thank you! Had the wrong private key in mind.

Reply:
I was thinking about the private key usage when tbot sends it back to the Teleport Cluster for signing.